src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java

```diff
@@ -3724,6 +3724,96 @@
     public static final int COMPACT_CTOR_DEF =
             GeneratedJavaTokenTypes.COMPACT_CTOR_DEF;
 
+    /**
+     * Beginning of a Java 14 Text Block literal,
+     * delimited by three double quotes.
+     *
+     * <p>For example:</p>
+     * <pre>
+     *         String hello = """
+     *                 Hello, world!
+     *                 """;
+     * </pre>
+     * <p>parses as:</p>
+     * <pre>
+     * |--VARIABLE_DEF
+     * |   |--MODIFIERS
+     * |   |--TYPE
+     * |   |   `--IDENT (String)
+     * |   |--IDENT (hello)
+     * |   |--ASSIGN (=)
+     * |   |   `--EXPR
+     * |   |       `--TEXT_BLOCK_LITERAL_BEGIN (""")
+     * |   |           |--TEXT_BLOCK_CONTENT (\n                Hello, world!\n                    )
+     * |   |           `--TEXT_BLOCK_LITERAL_END (""")
+     * |   `--SEMI (;)
+     * </pre>
+     *
+     * @since 8.36
+     */
+    public static final int TEXT_BLOCK_LITERAL_BEGIN =
+            GeneratedJavaTokenTypes.TEXT_BLOCK_LITERAL_BEGIN;
+
+    /**
+     * Content (text) of a Java 14 text block. This is a
+     * sequence of (possibly escaped) characters.
```

Comment:
Shouldn't we specify what is the escape character, which is `\`?

Reply:
Good point. Fixed.

Reply:
Note this escape goes directly to my other review item remaining. Maybe the error is in `possibly` escaped and should always be escaped.

Reply:
From JLS: _The use of the escape sequences `\"` and `\n` is permitted in a text block, but not necessary or recommended._

Reply:
We should mention that we substitude invisible '\n' (in linux single char symbol) with `\n` (2 chars).
By means of this we increase length of String object.

.... stop ...
Can we put in actual String symbol as is ? (single char)
is it a nuance of you AST printer that we need show content in single line and Printer is doing escapes ?
If this is a problems of AST printer - IT IS NOT a problem at all. Nobody care about our AST printer, nobody use it except for us in some cases.

Reply:
> Can we put in actual String symbol as is 
Do you mean in the AST?

Reply:
Also: https://github.com/checkstyle/checkstyle/blob/8e82c6cf0131875fa0a4f3912ad4169aa57ff5e1/src/main/java/com/puppycrawl/tools/checkstyle/AstTreeStringPrinter.java#L231

Reply:
Please prove that size of String object in AST that contains all values from text block is not bigger than in java runtime.

Reply:
The compiler removes all of the preceding whitespace, and we do not:

```
➜  text-blocks-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  text-blocks-grammar cat TestClass.java                             
public class TestClass {

	public static void main(String... args) {

		String s1 = """
                \\n""";
		String s2 = """
                \n""";
		String s3 = """
                \\""";

		String s4 = """
                \u000D""";
		String s5 = """
                \n""";
		String s6 = """
                \n
                  """;
		System.out.println("Length of s1: " + s1.length());
		System.out.println("Length of s2: " + s2.length());
		System.out.println("Length of s3: " + s3.length());
		System.out.println("Length of s4: " + s4.length());
		System.out.println("Length of s5: " + s5.length());
		System.out.println("Length of s6: " + s6.length());

	}
}
➜  text-blocks-grammar /usr/lib/jvm/java-14-openjdk/bin/java --enable-preview TestClass
Length of s1: 2
Length of s2: 1
Length of s3: 1
Length of s4: 1
Length of s5: 1
Length of s6: 2

```

Reply:
Consider this too:
```
  text-blocks-grammar /usr/lib/jvm/java-14-openjdk/bin/javac --enable-preview --source 14 TestClass.java
Note: TestClass.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
➜  text-blocks-grammar cat TestClass.java
public class TestClass {

	public static void main(String... args) {

String s1 = """
\\n""";
String s2 = """
\n""";
String s3 = """
\\""";

String s4 = """
\u000D""";
String s5 = """
\n""";
String s6 = """
\n
""";
	System.out.println("Length of s1: " + s1.length());
	System.out.println("Length of s2: " + s2.length());
	System.out.println("Length of s3: " + s3.length());
	System.out.println("Length of s4: " + s4.length());
	System.out.println("Length of s5: " + s5.length());
	System.out.println("Length of s6: " + s6.length());

	}
}
➜  text-blocks-grammar /usr/lib/jvm/java-14-openjdk/bin/java --enable-preview TestClass
Length of s1: 2
Length of s2: 1
Length of s3: 1
Length of s4: 1
Length of s5: 1
Length of s6: 2

```

```
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--METHOD_DEF -> METHOD_DEF [3:1]
    |   |--MODIFIERS -> MODIFIERS [3:1]
    |   |   |--LITERAL_PUBLIC -> public [3:1]
    |   |   `--LITERAL_STATIC -> static [3:8]
    |   |--TYPE -> TYPE [3:15]
    |   |   `--LITERAL_VOID -> void [3:15]
    |   |--IDENT -> main [3:20]
    |   |--LPAREN -> ( [3:24]
    |   |--PARAMETERS -> PARAMETERS [3:25]
    |   |   `--PARAMETER_DEF -> PARAMETER_DEF [3:25]
    |   |       |--MODIFIERS -> MODIFIERS [3:25]
    |   |       |--TYPE -> TYPE [3:25]
    |   |       |   `--IDENT -> String [3:25]
    |   |       |--ELLIPSIS -> ... [3:31]
    |   |       `--IDENT -> args [3:35]
    |   |--RPAREN -> ) [3:39]
    |   `--SLIST -> { [3:41]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [5:0]
    |       |   |--MODIFIERS -> MODIFIERS [5:0]
    |       |   |--TYPE -> TYPE [5:0]
    |       |   |   `--IDENT -> String [5:0]
    |       |   |--IDENT -> s1 [5:7]
    |       |   `--ASSIGN -> = [5:10]
    |       |       `--EXPR -> EXPR [5:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [5:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\\n [5:15]
result of node.getText().length: 4
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [6:3]
    |       |--SEMI -> ; [6:6]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [7:0]
    |       |   |--MODIFIERS -> MODIFIERS [7:0]
    |       |   |--TYPE -> TYPE [7:0]
    |       |   |   `--IDENT -> String [7:0]
    |       |   |--IDENT -> s2 [7:7]
    |       |   `--ASSIGN -> = [7:10]
    |       |       `--EXPR -> EXPR [7:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [7:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\n [7:15]
result of node.getText().length: 3
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [8:2]
    |       |--SEMI -> ; [8:5]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [9:0]
    |       |   |--MODIFIERS -> MODIFIERS [9:0]
    |       |   |--TYPE -> TYPE [9:0]
    |       |   |   `--IDENT -> String [9:0]
    |       |   |--IDENT -> s3 [9:7]
    |       |   `--ASSIGN -> = [9:10]
    |       |       `--EXPR -> EXPR [9:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [9:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\\ [9:15]
result of node.getText().length: 3
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [10:2]
    |       |--SEMI -> ; [10:5]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [12:0]
    |       |   |--MODIFIERS -> MODIFIERS [12:0]
    |       |   |--TYPE -> TYPE [12:0]
    |       |   |   `--IDENT -> String [12:0]
    |       |   |--IDENT -> s4 [12:7]
    |       |   `--ASSIGN -> = [12:10]
    |       |       `--EXPR -> EXPR [12:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [12:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\u000D [12:15]
result of node.getText().length: 7
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [13:6]
    |       |--SEMI -> ; [13:9]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [14:0]
    |       |   |--MODIFIERS -> MODIFIERS [14:0]
    |       |   |--TYPE -> TYPE [14:0]
    |       |   |   `--IDENT -> String [14:0]
    |       |   |--IDENT -> s5 [14:7]
    |       |   `--ASSIGN -> = [14:10]
    |       |       `--EXPR -> EXPR [14:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [14:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\n [14:15]
result of node.getText().length: 3
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [15:2]
    |       |--SEMI -> ; [15:5]
    |       |--VARIABLE_DEF -> VARIABLE_DEF [16:0]
    |       |   |--MODIFIERS -> MODIFIERS [16:0]
    |       |   |--TYPE -> TYPE [16:0]
    |       |   |   `--IDENT -> String [16:0]
    |       |   |--IDENT -> s6 [16:7]
    |       |   `--ASSIGN -> = [16:10]
    |       |       `--EXPR -> EXPR [16:12]
    |       |           `--TEXT_BLOCK_LITERAL_BEGIN -> """ [16:12]
    |       |               |--TEXT_BLOCK_CONTENT -> \n\n\n [16:15]
result of node.getText().length: 4
    |       |               `--TEXT_BLOCK_LITERAL_END -> """ [18:0]
......
```

Reply:
So, we aren't miscounting the actual line terminators, but our length is not accurate due to counting the whitespace preceding the text.

Reply:
It is good that keep spaces !!!!
We care about source file as text , not in what it result after compilation ( compiler remove a lot of stuff from source code, it is grey area). We might change this in future if there be bug demand.